grommet/grommet
# Remove defaultProps usages for functional components

#### What does this PR do?

Starting from React 19, `defaultProps` for functional components is deprecated. Starting from React 18.3, the corresponding error is thrown for a functional component with `defaultProps` defined.

Changes in this PR:
- replace `defaultProps` for functional components with default function parameters values;
- use `attrs` method of `styled-components` to pass theme to styled components which might be outside of theme context (required per https://github.com/grommet/grommet/issues/6741#issuecomment-1533232289);
- add custom hook to obtain theme from context with fallback to default value;
- configure eslint rule to not require `defaultProps` for functional components;
- update snapshots.

#### Where should the reviewer start?

No specific place to start.

#### What testing has been done on this PR?

- manual testing on storybook and real project using grommet;
- tests run result:
```
Test Suites: 97 passed, 97 total
Tests:       1548 passed, 1548 total
Snapshots:   1777 passed, 1777 total
```

#### How should this be manually tested?

Use React 18.3+ in an app. Check that styles are applied correctly and no errors about `defaultProps` are thrown into console.

#### Do Jest tests follow these best practices?


#### Any background context you want to provide?

[React 19 upgrade guide](https://react.dev/blog/2024/04/25/react-19-upgrade-guide)

#### What are the relevant issues?

Closes https://github.com/grommet/grommet/issues/6741

#### Screenshots (if appropriate)

N/A.

#### Do the grommet docs need to be updated?

Just the internal ones (if any).

#### Should this PR be mentioned in the release notes?

Yes.

#### Is this change backwards compatible or is it a breaking change?

Changes are backwards compatible.
